enhancements/machine-config/pin-and-pre-load-images.md

```diff
@@ -0,0 +1,346 @@
+configuration parameter to "", but that would affect all images, not just the
+pinned ones. This behavior needs to be changed in CRI-O so that pinned images
+aren't removed, regardless of the value of `version_file_persist`.
+
+The changes to pin the images will be done in a file inside the
+`/etc/crio/crio.conf.d` directory. To avoid potential conflicts with files
+manually created by the administrator the name of this file will be the name of
+the `PinnedImageSet` custom resource concatenated with the UUID assiged by the
+API server. For example, if the custom resource is this:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-pinned-images
+  uuid: 550a1d88-2976-4447-9fc7-b65e457a7f42
+spec:
+  pinnedImages:
+  - quay.io/openshift-release-dev/ocp-release@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+  ...
+```
+
+Then the complete path will be this:
+
+```txt
+/etc/crio/crio.conf.d/my-pinned-images-550a1d88-2976-4447-9fc7-b65e457a7f42.conf
+```
+
+The content of the file will the `pinned_images` parameter containing the list
+of images:
+
+```toml
+pinned_images=[
+  "quay.io/openshift-release-dev/ocp-release@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+  ...
+]
+```
+
+In addition to the list of images to be pinned, the `PinnedImageSet` custom
+resource will also contain a node selector. This is intended to support
+different sets of images for different kinds of nodes. For example, to pin
+different images for control plane and worker nodes the user could create two
+`PinnedImageSet` custom resources:
+
+```yaml
+# For control plane nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/control-plane: ""
+  pinnedImages:
+  ...
+
+---
+
+# For worker nodes:
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: PinnedImageSet
+metadata:
+  name: my-control-plane-pinned-images
+spec:
+  nodeSelector:
+    matchLabels:
+      node-role.kubernetes.io/worker: ""
+  pinnedImages:
+  ...
+```
+
+This is specially convenient for pinning images for upgrades: there are many
+images that are needed only by the control plane nodes and there is no need to
+have them consuming disk space in worker nodes.
+
+When no node selector is specified the images will be pinned in all the nodes
+of the cluster.
+
+When a `PinnedImageSet` custom resource is added, modified or deleted the
+machine config operator will create, modify or delete the configuration file,
+reload the CRI-O configuration (with the equivalent of `systemctl reload crio`)
+and then it will use the CRI-O gRPC API to run the equivalent of `crictl pull`
```

Comment:
pinned images don't currently get reset on reload, support for that will need to be added.

Reply:
Thanks @haircommander. Added a paragraph explaining that:

> Note that currently CRI-O doesn't reset pinned images on reload, support for
> that will need to be added.

Reply:
When will that be added?

What specifically do we mean by "reset" here? If I land a new PinnedImageSet config and reload will cri-o pick that up on reload but if I were to modify or delete it then those changes wouldn't be picked up on reload?

As long as new configs are applied on reload I think this is likely manageable, infact some customers may wish to preserve their ability to preform a z-stream rollback not wishing to delete the previous PIS until subsequent update. That would however mean that they need storage capacity for up to 3 PIS, old, current, and to-be applied versions. When applying the update they could remove the old version and allow those images to be garbage collected after reboot.